src/dinitctl.cc

```diff
@@ -1903,30 +1929,68 @@ static int enable_disable_service(int socknum, cpbuffer_t &rbuffer, service_dir_
     }
 
     if (socknum >= 0) {
+        if (verbose) {
+            cout << "Service '" << to << "' has been " << (enable ? "enabled" : "disabled") << "." << endl;
+        }
+
         // Check status of the service now
         auto m = membuf()
                 .append((char)cp_cmd::SERVICESTATUS)
                 .append(to_handle);
         write_all_x(socknum, m);
 
-        wait_for_reply(rbuffer, socknum);
+        int statussize = 6 + std::max(sizeof(pid_t), sizeof(int));;
+
+        // For an enable, we want to wait until the service has started so we can report any
+        // failure. But, if the service is already started, we won't get any service events, so we
+        // have to request status via SERVICESTATUS to catch that case. However, we may get
+        // a service event before the reply to SERVICESTATUS and in that case should use it to
+        // report status.
+        if (enable) {
+            int r = rbuffer.fill_to(socknum, 2);
+            while (r > 0 && rbuffer[0] >= 100) {
+                unsigned pktlen = (unsigned char) rbuffer[1];
+                fill_buffer_to(rbuffer, socknum, pktlen);
+
+                if (rbuffer[0] == (char)cp_info::SERVICEEVENT) {
+                    int ret = process_service_event(rbuffer, pktlen, to_handle, to, false /* start */,
+                            verbose);
+                    if (ret >= 0) {
+                        // Consume the outstanding reply packet before returning.
+                        wait_for_reply(rbuffer, socknum);
+                        if (rbuffer[0] != (char)cp_rply::SERVICESTATUS) {
+                            cerr << "dinitctl: protocol error." << endl;
+                            return 1;
+                        }
+                        fill_buffer_to(rbuffer, socknum, statussize + 2 /* 1 byte reserved */);
```

Comment:
The comment here: `/* 1 byte reserved */` - doesn't explain the `+ 2` in the code. Something like: `/* 1 byte packet type, 1 byte reserved */` would make more sense.

Reply:
it will be too ugly:
```
fill_buffer_to(rbuffer, socknum,
        statussize + 2 /* 1 byte packet type, 1 byte reserved */);
```
I wrote it in a separate line.